File: lld/strategy/strategy.go
package main

type PaymentMethod interface {
	Pay()
}

type PaymentProcessor struct {
	method PaymentMethod
}

func (pp PaymentProcessor) ProcessPayment() {
	pp.method.Pay()
}

func NewProcessPayment(pm PaymentMethod) PaymentProcessor {
	return PaymentProcessor{
		method: pm,
	}
}

type UPIPayment struct {
	name string
}

func NewUPIPayment(name string) UPIPayment {
	return UPIPayment{
		name: name,
	}
}

func (upi UPIPayment) Pay() {
	println(upi.name)
	println("This is UPI Payment")
}

type CardPayment struct {
	name string
}

func NewCardPayment(name string) CardPayment {
	return CardPayment{
		name: name,
	}
}

func (cc CardPayment) Pay() {
	println(cc.name)
	println("This is Card Payment")
}

func main() {
	pp := NewProcessPayment(NewUPIPayment("UPI"))
	pp.ProcessPayment()
	pp.method = NewCardPayment("Card")
	pp.ProcessPayment()
}

File: lld/strategy.js

class Payment {
    constructor(strategy) {
        this.strategy = strategy;
    }

    setStrategy(strategy) {
        this.strategy = strategy
    }

    processPayment() {
        console.log("process payment", this.strategy.name)
        this.strategy.pay() // so, each payment will mantain their own pay function
        // this is strategy pattern, which mean code should be open for extensions and close for 
        // modification
    }
}


// solid principle says, 
//1: S - single responsibility (each class should be seperate on it's own)
//2: O - open/closed resp (classed should be open for extension but close for modification)

class UPIPayment {
    name = "UPIPayment"
    pay() {
        console.log("This is UPI payment")
    }
}

class CardPayment {
    name = "CardPayment"
    pay() {
        console.log("This is card payment");
    }
}

// paypal option

payment = new Payment(new UPIPayment)
payment.processPayment()

payment.setStrategy(new CardPayment)
payment.processPayment()
